saml-metadata-bindings/src/main/java/uk/gov/ida/saml/metadata/CertificateChainValidationFilter.java

```
@@ -0,0 +1,178 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import javax.xml.namespace.QName;
+import java.security.KeyStore;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.HashSet;
+
+import static org.opensaml.xmlsec.keyinfo.KeyInfoSupport.getCertificates;
+
+public final class CertificateChainValidationFilter implements MetadataFilter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CertificateChainValidationFilter.class);
+
+    private final QName role;
+    private final CertificateChainValidator certificateChainValidator;
+    private final KeyStore keyStore;
+
+    public CertificateChainValidationFilter(
+        @NotNull final QName role,
+        @NotNull final CertificateChainValidator certificateChainValidator,
+        @NotNull final KeyStore keyStore) {
+
+        this.role = role;
+        this.certificateChainValidator = certificateChainValidator;
+        this.keyStore = keyStore;
+    }
+
+    public QName getRole() {
+        return role;
+    }
+
+    public CertificateChainValidator getCertificateChainValidator() {
+        return certificateChainValidator;
+    }
+
+    private KeyStore getKeyStore() {
+        return keyStore;
+    }
+
+    @Nullable
+    @Override
+    public XMLObject filter(@Nullable XMLObject metadata) {
+        if (metadata == null) {
+            return null;
+        }
+
+        try {
+            if (metadata instanceof EntityDescriptor) {
+                processEntityDescriptor((EntityDescriptor) metadata);
+            } else if (metadata instanceof EntitiesDescriptor) {
+                processEntityGroup((EntitiesDescriptor) metadata);
+            } else {
+                LOG.error("Internal error, metadata object was of an unsupported type: {}", metadata.getClass().getName());
+            }
+        } catch (Throwable t) {
```

Do we need to catch Throwable?  It seems the only exception we'll get here is CertificateConversionException (which wraps CertificateException) - or some sort of serious JVM error.  Do we want to swallow all JVM errors in our filter?